staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator_test.go

```diff
@@ -172,6 +175,99 @@ func BenchmarkKeyFunc(b *testing.B) {
 	})
 }
 
+func TestSharedLookup(t *testing.T) {
+	t.Run("actually shared", func(t *testing.T) {
+		var lookups uint32
+		c := make(chan struct{})
+		a := New(authenticator.TokenFunc(func(ctx context.Context, token string) (*authenticator.Response, bool, error) {
+			<-c
+			atomic.AddUint32(&lookups, 1)
+			return &authenticator.Response{User: &user.DefaultInfo{Name: "chewbaca"}}, true, nil
+		}), true, time.Minute, 0)
+
+		var wg sync.WaitGroup
+		for i := 0; i < 10; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				a.AuthenticateToken(context.Background(), "")
+			}()
+		}
+		// no good way to make sure that all the callers are queued so we sleep.
+		time.Sleep(10 * time.Millisecond)
+		close(c)
+		wg.Wait()
+
+		if lookups != 1 {
+			t.Fatalf("unexpected number of lookups: got=%d, want=1", lookups)
+		}
+	})
+
+	t.Run("first caller bails, second caller gets result", func(t *testing.T) {
+		c := make(chan struct{})
+		a := New(authenticator.TokenFunc(func(ctx context.Context, token string) (*authenticator.Response, bool, error) {
+			<-c
+			return &authenticator.Response{User: &user.DefaultInfo{Name: "chewbaca"}}, true, nil
+		}), true, time.Minute, 0)
+
+		var wg sync.WaitGroup
+		wg.Add(2)
+
+		ctx1, cancel1 := context.WithCancel(context.Background())
+		go func() {
+			defer wg.Done()
+			a.AuthenticateToken(ctx1, "")
+		}()
+
+		ctx2 := context.Background()
+
+		var err error
+		go func() {
+			defer wg.Done()
+			_, _, err = a.AuthenticateToken(ctx2, "")
+		}()
+
+		time.Sleep(10 * time.Millisecond)
+		cancel1()
+		close(c)
+		wg.Wait()
+
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
```

Comment:
check the response as well?

Reply:
Done. Also note that the other tests in this file are covering this as well.